multidevice_discussion.py

```diff
@@ -0,0 +1,197 @@
+
+
+#########
+# Case A: A multidevice task launches a multi-device application. Task body is run a single time.
+#########
+
+# Comment(wlr): Case A is what I thought our MultiDevice semantics are
+
+def worker_run():
+    # Pseudo-code of Parla internals
+
+    # Get the current task
+    task = get_current_task()
+
+    # Get the current device list
+    active_devices = get_active_devices()
+    stream_list = [device.get_stream() for device in active_devices]
+
+    with create_environment(active_devices, stream_list) as e:
+        # Run the task body
+        task.body(e.locals)
+        # Record events on the streams
+        events = [stream.record_event() for stream in stream_list]
+        # Pass events to dependents
+        task.notify_dependents(events)
+
+    #Implicit synchronization on all streams at the end of the context manager
+
+
+########
+# Case A1: A multidevice task launches a library multi-device application
+########
+
+
+@spawn(gpus=2)
+def multidevice_application(locals):
+
+    # Get real device ids from local devices
+    device_ids = [device.id for device in locals.devices]
+    streams = [device.stream for device in locals.devices]
+
+    # Launch a library kernel on the real devices
+    # Assume A,B already distributed and in scope
+    NVIDIA_GEMM_MG[device_ids, streams](A, B)
+
+
+########
+# Case A2: A multidevice task launches a user written multi-device application
+########
+
+
+@spawn(gpus=2)
+def multidevice_application(locals):
+
+    # Assume default is equipartitioning
+    A = Crosspy(10, [locals.devices[0], locals.devices[1]])
+    B = Crosspy(10, [locals.devices[0], locals.devices[1]])
+
+    # Launch a user written application on the real devices
+    with locals.devices[0]:
```

Comment:
This is shorthand for:

```python
with cp.cuda.Device(locals.device[0].idx):
    with locals.device[0].stream:
        // Do stuff
```

Reply:
Now I see why you said we don't need the outer loop in the runtime. This is great. But I think we may need to add one more parameter for better programming model. In some cases, it could complicate the programming model due to `with cp.cuda.Device(device_idx):` from `with locals.devices[0]`.

For example, let's assume that we have the case that runs SIMD.

```
@spawn(gpus=2)
def test():
  beg = locals.current_device == 0? 0 : index_range[locals.current_device - 1]
  end = index_range[locals.current_device]
  # computation kernels
  kernel(a[beg:end], b[beg:end])..
```
The above is a frequent pattern for parallel progarmming on multi-host executions.
In this case, we don't need to diverge the body.
I think users don't want to consider and write
```
with locals.devices[0], locals.devices[1]:
```
but can expect that the runtime just dispatches this body to the 2 GPUs.
(So to speak, expect that the runtime calls with decorators at the runtime level)

Maybe we can have one more param like this

```
@spawn(gpus=2, disable_context_dispatching=true) # whatever, to enable your example
```
Or
we can maybe modify the semantic of `with locals.devices[0]` to just 
```
if locals.devices[0] == locals.current_context_device:
```
and the runtime performs the original context that:

```
with cp.cuda.Device(locals.device[0].idx):
  with locals.device[0].stream: // Do stuff
```

Then your example and my example both don't need additional parameter with the same semantic.

Reply:
My pushback on the MPI-style SIMD semantics comes from the following reasons.
Not all of these are of equal importance, I'm just collecting my thoughts. 

1. It cannot support calls to external multigpu libraries and functions. I view this as the most vital usecase & motivator for multigpu tasks. CrossPy itself is a good example of a multi-device library; it has communication collectives that would not want to be replicated for each device by repeating the body.

```python

A = CrossPy(10, [gpu(0), gpu(1)])
B = CrossPy(10, [gpu(0), gpu(1)])

@spawn(gpus=2)
def test(locals):
    C = A + B

```

iiuc, the above would break in the Case B semantics. 

2. It breaks sequential semantics. 

3. Without a collection internal synchronization primitives, I don't see a clear advantage over a loop of single-device tasks. 

```python

for i in range(NUM_DEVICES):
    @spawn(placement=devices[i])
    def test(T[i], locals):
        beg = locals.current_device == 0? 0 : index_range[locals.current_device - 1]
        end = index_range[locals.current_device]

        # computation kernels
        kernel(a[beg:end], b[beg:end])
```
Benefits of single-device tasks: 
- Maintain sequential semantics
- Each task runs concurrently with multithreading on the device set. Users don't have to worry as much about which CuPy calls are synchronous and which are asynchronous.


```python

@spawn(T[0], gpus=2)
def test(locals):
  beg = locals.current_device == 0? 0 : index_range[locals.current_device - 1]
  end = index_range[locals.current_device]
  # computation kernels
  kernel(a[beg:end], b[beg:end])
```

Benefits of multi-device dispatch:
- Less tasks in the runtime. 
- Easier to specify dependencies on the whole collection of kernels

In the case of multidevice tasks, for array types I don't see anything that couldn't be done with a `map` syntax wrapper on top of CrossPy. 

```python

@spawn(T[0], gpus=2)
def test(locals):
    locals.map(kernel, (a, index_range), (b index_range))
```

4. CuPy's API is very unclear about which calls are synchronous and which are asynchronous. Single thread context dispatch hides this loop from the user and makes it easier to accidentally write a multigpu task that is not concurrent. On the other hand, multiple thread context dispatch seems to add a significant amount of complexity to the runtime.

5. The implementation of internal synchronization primitives on subsets of the devices through coroutines seems fairly nasty if I'm understanding how they would need to be written correctly.

Reply:
Before demonstrate my opinions on your thoughts, first please clarify those terms:
1) single-device tasks: does it mean a multi-device task replicating the same body across devices transparently?
2) multi-device dispatch: what do you mean? are you referring to the same case? I am asking b/c both examples dispatch code snippets onto multiple devices.
3) what do you mean single thread context? both are using a single thread context. 

(but I can see what you are talking about. to summarize, there are multple pros/cons from those two cases and we may need to choose one of them. I think for this, we should refine and list what you wrote and discuss it with profs, as you said. And your 1st reason changed my mind. I now prefer your suggestion. But I also want to point out that your pushback is not pure MPI style since MPI does not change any host context at application level. All same codes are run on the entire hosts. Each host performs proper codes by checking MPI rank like the case B, the second example)

Reply:
1. By `single-device task` I mean a task whose assigned placement is a set containing only 1 device. It reserves resources on only 1 device and runs within a single device context. (These are 'classical' Parla tasks)

2. I was trying to use your term `context_dispatching` from the above comment to refer to splitting the task into multiple execution passes of the task body on different contexts. 

By "Single Thread Context Dispatch" I mean https://github.com/ut-parla/parla-experimental/pull/10/files#diff-b4556241a1419472759c0bb002fef0c9d90f6bea30e70dccd926ba1940831413R95

By "Multiple Thread Context Dispatch" I mean https://github.com/ut-parla/parla-experimental/pull/10/files#diff-b4556241a1419472759c0bb002fef0c9d90f6bea30e70dccd926ba1940831413R123

I agree they're not very clear terms. 

3. I think clarifying 2 answers this. Let me know if it doesn't?


(By MPI-Style I mean "Case B" examples)